File: app/src/main/java/ua/in/quireg/anothermovieapp/common/UriHelper.java
package ua.in.quireg.anothermovieapp.common;


import android.net.Uri;

import ua.in.quireg.anothermovieapp.BuildConfig;

public class UriHelper {
    private static final String LOG_TAG = UriHelper.class.getSimpleName();
    private static final String API_KEY = BuildConfig.MOVIE_DATABASE_API_KEY;

    public static Uri getMoviesListUri(String tag){
        MovieAppLogger.v(LOG_TAG, "getMoviesListUri()");
        switch (tag){
            case Constants.POPULAR:
                return new Uri.Builder()
                        .scheme("https")
                        .authority("api.themoviedb.org")
                        .appendPath("3")
                        .appendPath("movie")
                        .appendPath("popular")
                        .appendQueryParameter("api_key", API_KEY)
                        .build();
            case Constants.TOP_RATED:
                return new Uri.Builder()
                        .scheme("https")
                        .authority("api.themoviedb.org")
                        .appendPath("3")
                        .appendPath("movie")
                        .appendPath("top_rated")
                        .appendQueryParameter("api_key", API_KEY)
                        .build();
        }
        return null;
    }

    public static Uri getMoviesListPageUri(String tag, String page){
        MovieAppLogger.v(LOG_TAG, "getMoviesListPageUri()");

        switch (tag){
            case Constants.POPULAR:
                return new Uri.Builder()
                        .scheme("https")
                        .authority("api.themoviedb.org")
                        .appendPath("3")
                        .appendPath("movie")
                        .appendPath("popular")
                        .appendQueryParameter("page", page)
                        .appendQueryParameter("api_key", API_KEY)
                        .build();
            case Constants.TOP_RATED:
                return new Uri.Builder()
                        .scheme("https")
                        .authority("api.themoviedb.org")
                        .appendPath("3")
                        .appendPath("movie")
                        .appendPath("top_rated")
                        .appendQueryParameter("page", page)
                        .appendQueryParameter("api_key", API_KEY)
                        .build();
        }
        return null;
    }

    public static Uri getImageUri(String imagePath, String imageSize){
        MovieAppLogger.v(LOG_TAG, "getImageUri()");
        return new Uri.Builder()
                .scheme("https")
                .authority("image.tmdb.org")
                .appendPath("t")
                .appendPath("p")
                .appendPath(imageSize)
                .appendPath(imagePath)
                .build()
                .normalizeScheme();
    }

    public static Uri getMovieUriById(String id){
        MovieAppLogger.v(LOG_TAG, "getMovieUriById()");
        return new Uri.Builder()
                .scheme("https")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath(id)
                .appendQueryParameter("api_key", API_KEY)
                .build();
    }

    public static Uri getMovieTrailerUriById(String id){
        MovieAppLogger.v(LOG_TAG, "getMovieTrailerUriById()");
        return new Uri.Builder()
                .scheme("https")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath(id)
                .appendPath("videos")
                .appendQueryParameter("api_key", API_KEY)
                .build();
    }

    public static Uri getMovieReviewsUri(String id){
        MovieAppLogger.v(LOG_TAG, "getMovieReviewsUri()");
        return new Uri.Builder()
                .scheme("https")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath(id)
                .appendPath("reviews")
                .appendQueryParameter("api_key", API_KEY)
                .build();
    }
    public static Uri getMovieReviewsUriForPage(String id, String page){
        MovieAppLogger.v(LOG_TAG, "getMovieReviewsUri()");
        return new Uri.Builder()
                .scheme("https")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath(id)
                .appendPath("reviews")
                .appendQueryParameter("page", page)
                .appendQueryParameter("api_key", API_KEY)
                .build();
    }

    public static Uri getYouTubeLinkToPlay(String id){
        MovieAppLogger.v(LOG_TAG, "getYouTubeLinkToPlay()");
        return new Uri.Builder()
                .scheme("https")
                .authority("youtube.com")
                .appendPath("watch")
                .appendQueryParameter("v", id)
                .build();
    }

    public static Uri getYouTubeTrailerPreviewLink(String trailer_id){
        MovieAppLogger.v(LOG_TAG, "getYouTubeTrailerPreviewLink()");
        return new Uri.Builder()
                .scheme("https")
                .authority("i1.ytimg.com")
                .appendPath("vi")
                .appendPath(trailer_id)
                .appendPath("hqdefault.jpg")
                .build();
    }

}

File: app/src/main/java/com/anothermovieapp/common/UriHelper.kt
package com.anothermovieapp.common

import android.net.Uri
import com.anothermovieapp.BuildConfig

object UriHelper {
    private val LOG_TAG = UriHelper::class.java.simpleName
    private const val API_KEY = BuildConfig.MOVIE_DATABASE_API_KEY
    fun getMoviesListUri(tag: String?): Uri? {
        MovieAppLogger.v(LOG_TAG, "getMoviesListUri()")
        when (tag) {
            Constants.POPULAR -> return Uri.Builder()
                    .scheme("https")
                    .authority("api.themoviedb.org")
                    .appendPath("3")
                    .appendPath("movie")
                    .appendPath("popular")
                    .appendQueryParameter("api_key", API_KEY)
                    .build()
            Constants.TOP_RATED -> return Uri.Builder()
                    .scheme("https")
                    .authority("api.themoviedb.org")
                    .appendPath("3")
                    .appendPath("movie")
                    .appendPath("top_rated")
                    .appendQueryParameter("api_key", API_KEY)
                    .build()
        }
        return null
    }

    fun getMoviesListPageUri(tag: String?, page: String?): Uri {
        MovieAppLogger.v(LOG_TAG, "getMoviesListPageUri()")
        return when (tag) {
            Constants.POPULAR -> Uri.Builder()
                    .scheme("https")
                    .authority("api.themoviedb.org")
                    .appendPath("3")
                    .appendPath("movie")
                    .appendPath("popular")
                    .appendQueryParameter("page", page)
                    .appendQueryParameter("api_key", API_KEY)
                    .build()
            Constants.TOP_RATED -> Uri.Builder()
                    .scheme("https")
                    .authority("api.themoviedb.org")
                    .appendPath("3")
                    .appendPath("movie")
                    .appendPath("top_rated")
                    .appendQueryParameter("page", page)
                    .appendQueryParameter("api_key", API_KEY)
                    .build()
            else -> Uri.Builder()
                    .scheme("https")
                    .authority("api.themoviedb.org")
                    .appendPath("3")
                    .appendPath("movie")
                    .appendPath("top_rated")
                    .appendQueryParameter("page", page)
                    .appendQueryParameter("api_key", API_KEY)
                    .build()
        }
    }

    fun getImageUri(imagePath: String?, imageSize: String?): Uri {
        MovieAppLogger.v(LOG_TAG, "getImageUri()")
        return Uri.Builder()
                .scheme("https")
                .authority("image.tmdb.org")
                .appendPath("t")
                .appendPath("p")
                .appendPath(imageSize)
                .appendPath(imagePath)
                .build()
                .normalizeScheme()
    }

    fun getMovieUriById(id: String?): Uri {
        MovieAppLogger.v(LOG_TAG, "getMovieUriById()")
        return Uri.Builder()
                .scheme("https")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath(id)
                .appendQueryParameter("api_key", API_KEY)
                .build()
    }

    fun getMovieTrailerUriById(id: String?): Uri {
        MovieAppLogger.v(LOG_TAG, "getMovieTrailerUriById()")
        return Uri.Builder()
                .scheme("https")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath(id)
                .appendPath("videos")
                .appendQueryParameter("api_key", API_KEY)
                .build()
    }

    fun getMovieReviewsUri(id: String?): Uri {
        MovieAppLogger.v(LOG_TAG, "getMovieReviewsUri()")
        return Uri.Builder()
                .scheme("https")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath(id)
                .appendPath("reviews")
                .appendQueryParameter("api_key", API_KEY)
                .build()
    }

    fun getMovieReviewsUriForPage(id: String?, page: String?): Uri {
        MovieAppLogger.v(LOG_TAG, "getMovieReviewsUri()")
        return Uri.Builder()
                .scheme("https")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath(id)
                .appendPath("reviews")
                .appendQueryParameter("page", page)
                .appendQueryParameter("api_key", API_KEY)
                .build()
    }

    fun getYouTubeLinkToPlay(id: String?): Uri {
        MovieAppLogger.v(LOG_TAG, "getYouTubeLinkToPlay()")
        return Uri.Builder()
                .scheme("https")
                .authority("youtube.com")
                .appendPath("watch")
                .appendQueryParameter("v", id)
                .build()
    }

    fun getYouTubeTrailerPreviewLink(trailer_id: String?): Uri {
        MovieAppLogger.v(LOG_TAG, "getYouTubeTrailerPreviewLink()")
        return Uri.Builder()
                .scheme("https")
                .authority("i1.ytimg.com")
                .appendPath("vi")
                .appendPath(trailer_id)
                .appendPath("hqdefault.jpg")
                .build()
    }
}